### PR TITLE
Add name constraints to CA

### LINKED ISSUE #11075: Is the Certificate Authority name-constrained somehow?
I wonder if the Certificate Authority we have to add to root stores to work with web-platform-tests is name-constrained somehow or if an attacker can use it mint certificates.

### PR DESCRIPTION
**Pull request by franziskuskiefer**
Something like this should fix issue #11075 (fixes #11075).
I don't know how to run this so I can't tell for sure. But adding something of the form `nameConstraints=permitted;DNS:.example.com,permitted;DNS:.example.org` will give you what you want.

### CODE REVIEW

**Comment by annevk** [comment]
Thanks @franziskuskiefer!

@jgraham @gsnedders @jugglinmike would you be willing to iterate on this further?

File: tools/wptserve/wptserve/sslutils/openssl.py [hunk 1]
```diff
@@ -213,6 +218,7 @@ def get_config(root_dir, hosts, duration=30):
 subjectKeyIdentifier=hash
 authorityKeyIdentifier=keyid:always,issuer:always
 keyUsage = keyCertSign
+%(constraints_line)s
```
Comment (mozfreddyb) [inline tools/wptserve/wptserve/sslutils/openssl.py]: Drive-by comment: This line above uses a variable that is local to the formatted string, which needs to be defined when formatting the string. This is named as the same variable as in the outer code block, but not implicitly linked.
To make this work, the in-line dictionary after the formatted string would need an additional key/value pair of

`"constraints_line": constraints_line,`

**Comment by foolip** [comment]
@gsnedders, I assigned to you since https://github.com/web-platform-tests/wpt/issues/11075 is on our list of important infra issues.

**Comment by jugglinmike** [comment]
We recently received [a feature request to support arbitrary subdomains](https://github.com/web-platform-tests/wpt/issues/13465). We're not doing that (yet), but it might be wise to account for the possibility in this patch by relaxing the constraints a bit. As far as I know (which is not very far), it will be no less secure to constrain to:

- `web-platform.test`
- `.web-platform.test`
- `not-web-platform.test`
- `.not-web-platform.test`

Though now I'm wondering if we could simply constrain to `.test`.

**Comment by gsnedders** [comment]
@jugglinmike 

> Though now I'm wondering if we could simply constrain to `.test`.

I'd strongly oppose this, given I (and probably others) have other local dev environments served on `.test`.

**Comment by jgraham** [comment]
I don't think the arbitary subdomains thing needs to be considered here; if we change the range of supported hosts it will in any case be necessary to update the SAN line on the certs so there is barely any advantage in preempting that change for the constraints line only.

**Comment by foolip** [comment]
Ping @gsnedders for review.

File: tools/wptserve/wptserve/sslutils/openssl.py [hunk 2]
```diff
@@ -114,16 +114,18 @@ def make_subject(common_name,
...
+def make_name_constraints(hosts):
+    return ",".join("permitted;DNS:%s" % host for host in hosts)
 
 def get_config(root_dir, hosts, duration=30):
     if hosts is None:
         san_line = ""
+        constraints_line = ""
     else:
         san_line = "subjectAltName = %s" % make_alt_names(hosts)
+        constraints_line = "nameConstraints = " + make_name_constraints(hosts)
```
Comment (gsnedders) [inline tools/wptserve/wptserve/sslutils/openssl.py]: This will lead to trying to put `nameConstraints` on the server cert, rather than the CA cert.

From `_generate_ca` we call `get_config` with `hosts=None`:

```
  web-platform-tests/tools/wptserve/wptserve/sslutils/openssl.py(399)_generate_host_cert()
-> self._generate_ca()
 web-platform-tests/tools/wptserve/wptserve/sslutils/openssl.py(342)_generate_ca()
-> with self._config_openssl(None) as openssl:
  web-platform-tests/tools/wptserve/wptserve/sslutils/openssl.py(43)__enter__()
-> f.write(get_config(self.base_path, self.hosts, self.duration))
```

(interestingly `pdb` seems to be omitting `_config_openssl` in that!)

As such, this doesn't actually fix the underlying issue! I think we need to have a `is_ca` arg here?

**Review by gsnedders — CHANGES_REQUESTED** [review]
@foolip I actually started on this yesterday, but couldn't figure out why it wasn't working. :)

@jgraham @franziskuskiefer This PR should probably also update `tools/certs`, preferably also providing some documentation on how to update them!

**Comment by gsnedders** [comment]
@jgraham what do I need to do to update `tools/certs`? I presume I also need to make it have longer validity?

**Review by jgraham — APPROVED** [review]

**Comment by jgraham** [comment]
Updating the certs is a bit of a black art at the moment. You need to change the wptserve config to use openssl as the config type and add a duration key to the openssl part of the config with the desired duration. Then if you run the server you can copy the generated files out of the temporary location and into the irectory containing the pregenerated certs.

**Comment by foolip** [comment]
This PR is blocked on the required "Travis CI - Pull Request" check after https://github.com/web-platform-tests/wpt/issues/14499. In order to trigger it, I will close and reopen this PR.

**Review by gsnedders — APPROVED** [review]

File: tools/certs/.gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+/0*.pem
```
Comment (jgraham) [inline tools/certs/.gitignore]: Add this to the root .gitignore please (some tooling we have only works with the root).

**Review by jgraham — APPROVED** [review]
Subject to .gitignore change.

**Review by jgraham — APPROVED** [review]
